### PR TITLE
Add support for multiple chain types in the configuration

### PR DESCRIPTION
**Pull request by romac**
Closes: #2240

## Description

See the proposal on the linked issue.

______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] Added tests: integration (for Hermes) or unit/mock tests (for modules). 
- [x] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: relayer-cli/src/cli_utils.rs [hunk 1]
```diff
@@ -55,20 +56,13 @@ pub fn spawn_chain_runtime(config: &Config, chain_id: &ChainId) -> Result<impl C
     spawn_chain_runtime_generic::<BaseChainHandle>(config, chain_id)
 }
 
-pub fn spawn_chain_runtime_generic<Chain: ChainHandle>(
+#[allow(unreachable_patterns)]
```
Comment (ancazamfir) [inline relayer-cli/src/cli_utils.rs:59]: remove?
```suggestion
```

**Comment by adizere** [comment]
> Can you also add some comment in `config.toml` and maybe show the default value:
> 
> ```
> type = 'CosmosSdk'
> ```

I'm wondering if we should document this chain type feature already in config.toml (and release it with v1), since it's rather experimental.

**Comment by ancazamfir** [comment]
> > Can you also add some comment in `config.toml` and maybe show the default value:
> > ```
> > type = 'CosmosSdk'
> > ```
> 
> I'm wondering if we should document this chain type feature already in config.toml (and release it with v1), since it's rather experimental.

@romac had the same comment offline. I am ok either way. I expect this to be used when other chain types will be supported, e.g. substrate. Not sure at what stage substrate integration is now was wondering if it would be useful for them to upstream these changes.

**Comment by adizere** [comment]
> Not sure at what stage substrate integration is now, was wondering if it would be useful for them to upstream these changes.

Good point! Another use-case for the chain type feature is integration with basecoin-rs.

From what I can tell, we're signalling the changes here via (doc)comments and the public API here
https://github.com/informalsystems/ibc-rs/blob/22652ee2aea75cd559aedd151813dbadadd4048b/relayer/src/chain.rs#L17

Not sure how to make this API better for substrate or other relayer developers, but as a first approach this looks good to me! If this proves to be too obscure, then I agree we should highlight it in config.toml.

File: relayer/src/upgrade_chain.rs [hunk 2]
```diff
@@ -66,12 +66,12 @@ pub struct UpgradePlanOptions {
...
 pub fn build_and_send_ibc_upgrade_proposal(
-    mut dst_chain: CosmosSdkChain, // the chain which will undergo an upgrade
-    src_chain: CosmosSdkChain, // the source chain; supplies a client state for building the upgrade plan
+    dst_chain: impl ChainHandle, // the chain which will undergo an upgrade
+    src_chain: impl ChainHandle, // the source chain; supplies a client state for building the upgrade plan
     opts: &UpgradePlanOptions,
 ) -> Result<TxHash, UpgradeChainError> {
     let upgrade_height = dst_chain
-        .query_chain_latest_height()
+        .query_latest_height() // FIXME(romac): Use query_chain_latest_height once added to ChainHandle
```
[the remark this answers is not in the document]
Comment (adizere) [reply relayer/src/upgrade_chain.rs:74]: I agree, we should tackle this with issue 2185!

**Review by adizere — APPROVED** [review]
Looks good, great work Romain!

Pinging @plafer here because the present PR with #2226 will likely conflict, so be aware of the order of merging.

**Comment by mzabaluev** [comment]
On the third hand, deserialization of the chain type could just be case-insensitive. It should be if this format is also used in CLI, because 1) CLI users on Unix traditionally hate using upper case unless necessary; 2) CLI users on Windows are used to parameters being case-insensitive.

File: relayer/src/upgrade_chain.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (romac) [reply relayer/src/upgrade_chain.rs:74]: I can also issue the right call here and add the appropriate method to the handle or would you rather we do this in a follow-up PR?

**Comment by romac** [comment]
> On the third hand, deserialization of the chain type could just be case-insensitive.

Agreed. I'll see if I can make it both case insensitive and ignore dashes as well.

**Comment by romac** [comment]
> > On the third hand, deserialization of the chain type could just be case-insensitive.
> 
> Agreed. I'll see if I can make it both case insensitive and ignore dashes as well.

Done here: https://github.com/informalsystems/ibc-rs/pull/2228/commits/111f3f0a4146c72f83c4a40916b95941ee14b46b

File: relayer/src/upgrade_chain.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ancazamfir) [reply relayer/src/upgrade_chain.rs:74]: let's do it in a follow up PR

**Pull request merged**